javascript/ql/src/Security/CWE-730/ServerCrash.ql

```diff
@@ -0,0 +1,189 @@
+) {
+  async.getAsyncCallee() = cb and
+  rh.getAstNode() = invokesCallbackThatThrowsUncaughtException(async, thrower)
+}
+
+/**
+ * A call that may cause a function to be invoked in an asynchronous context outside of the visible source code.
+ */
+class AsyncSentinelCall extends DataFlow::CallNode {
+  Function asyncCallee;
+
+  AsyncSentinelCall() {
+    exists(DataFlow::FunctionNode node | node.getAstNode() = asyncCallee |
+      // manual models
+      exists(string memberName |
+        not "Sync" = memberName.suffix(memberName.length() - 4) and
+        this = NodeJSLib::FS::moduleMember(memberName).getACall() and
+        node = this.getCallback([1 .. 2])
+      )
+      // (add additional cases here to improve the query)
+    )
+  }
+
+  /**
+   * Gets the callee that is invoked in an asynchronous context.
+   */
+  Function getAsyncCallee() { result = asyncCallee }
+}
+
+/**
+ * A callback provided to an asynchronous call (heuristic).
+ */
+class AsyncCallback extends Function {
+  AsyncCallback() { any(AsyncSentinelCall c).getAsyncCallee() = this }
+}
+
+/**
+ * A node that is likely to throw an exception.
+ *
+ * This is the primary extension point for this query.
+ */
+abstract class LikelyExceptionThrower extends ASTNode { }
+
+/**
+ * A `throw` statement.
+ */
+class TrivialThrowStatement extends LikelyExceptionThrower, ThrowStmt { }
+
+/**
+ * Empty class for avoiding emptiness checks from the compiler when there are no Expr-typed instances of the LikelyExceptionThrower type.
+ */
+class CompilerConfusingExceptionThrower extends LikelyExceptionThrower {
+  CompilerConfusingExceptionThrower() { none() }
+}
+
+/**
+ * Edges that builds an explanatory graph that follows the mental model of how the the exception flows.
+ *
+ * - step 1. exception is thrown
+ * - step 2. exception escapes the enclosing function
+ * - step 3. exception follows the call graph backwards until an async callee is encountered
+ * - step 4. (at this point, the program crashes)
+ */
+query predicate edges(ASTNode pred, ASTNode succ) {
+  exists(LikelyExceptionThrower thrower | main(_, _, _, thrower) |
+    pred = thrower and
+    succ = thrower.getContainer()
+    or
+    exists(DataFlow::InvokeNode invk, Function fun |
+      fun = throwsUncaughtExceptionInAsyncContext(thrower)
+    |
+      succ = invk.getAstNode() and
+      pred = invk.getACallee() and
+      pred = fun
+      or
+      succ = fun and
+      succ = invk.getContainer() and
+      pred = invk.getAstNode()
+    )
+  )
+}
+
+/**
+ * A node in the `edge/2` relation above.
+ */
+query predicate nodes(ASTNode node) {
+  edges(node, _) or
+  edges(_, node)
+}
+
+from
+  HTTP::RouteHandler rh, AsyncSentinelCall async, DataFlow::Node callbackArg, AsyncCallback cb,
+  ExprOrStmt crasher
+where
+  main(rh, async, cb, crasher) and
+  callbackArg.getALocalSource().getAstNode() = cb and
+  async.getAnArgument() = callbackArg
```

Comment:
(Optional). 
If you implement a `getAsyncCallArgument()` in `AsyncSentinelCall`, then you can make this where clause a bit prettier.  

```suggestion
  callbackArg = async.getAsyncCallArgument()
```

An implementation of `AsyncSentinelCall` could be: 

```CodeQL
class AsyncSentinelCall extends DataFlow::CallNode {
  int arg;

  AsyncSentinelCall() {
    exists(this.getCallback(arg)) and
    // manual models
    exists(string memberName |
      not "Sync" = memberName.suffix(memberName.length() - 4) and
      this = NodeJSLib::FS::moduleMember(memberName).getACall() and
      arg = [1 .. 2]
    )
    // (add additional cases here to improve the query)
  }

  /**
   * Gets the callee that is invoked in an asynchronous context.
   */
  Function getAsyncCallee() { result = this.getCallback(arg).getFunction() }

  DataFlow::Node getAsyncCallArgument() { result = this.getArgument(arg) }
}
```

Reply:
`let callbackArg1 = cb1; callbackArg2 = cb2; fs.someFunction(x, callbackArg1, callbackArg2)` could then result in a message with a `callbackArg1`/`cb2` pair, in theory at least.

Reply:
Not if you use a `int arg` field to synchronize the two, right?

Reply:
`arg` does not synchronize the results of `getAsyncCallee` and `getAsyncCallArgument`. There is only one entity for `AsyncSentinellCall async` in my example, and it has `arg = 1 or arg = 2` internally.